### PR TITLE
Cluster upgrade test

### LINKED ISSUE #621: [E2E] Cluster upgrade test
**Describe the solution you'd like**
We can implement an e2e test to verify cluster upgrade in BYOH.

**Environment:**

- Cluster-api-provider-bringyourownhost version: 
- Kubernetes version: (use `kubectl version --short`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by Madhur97**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #621 

**Additional information**

**Special notes for your reviewer**

### CODE REVIEW

File: test/e2e/cluster_upgrade_test.go [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+		cancelWatches                context.CancelFunc
+		clusterResources             *clusterctl.ApplyClusterTemplateAndWaitResult
+		dockerClient                 *client.Client
+		byoHostCapacityPool          = 4
+		byoHostName                  string
+		allbyohostContainerIDs       []string
+		allAgentLogFiles             []string
+		coreDNSUpgradeVersion        = "v1.8.6"
+		kubernetesVersionUpgradeTo   = "v1.23.5"
+		kubernetesVersionUpgradeFrom = "v1.23.4"
```
Comment (anusha94) [inline test/e2e/cluster_upgrade_test.go:38]: Let's try a minor version upgrade? We already should have bundles available for `v1.22.x`
So let's change the test to upgrade from `v1.22.x` to `v1.23.y`

File: .github/workflows/e2e.yml [hunk 2]
```diff
@@ -25,5 +25,8 @@ jobs:
     - name: Set netfilter conntrack max
       run: sudo sysctl -w net.netfilter.nf_conntrack_max=131072
 
+    - name: Run K8s upgrade e2e tests
```
Comment (anusha94) [inline .github/workflows/e2e.yml:28]: We don't want to run upgrade test with every PR commit - it should be run as part of run-all-suite (#584 ). Once we are sure this test passes, let's move it along with the rest of e2e tests workflow.

File: test/e2e/cluster_upgrade_test.go (hunk 1 above)
Comment (Madhur97) [reply test/e2e/cluster_upgrade_test.go]: Tested with minor version upgrade (v1.22.3 => v1.23.5) , it is working fine.

File: .github/workflows/e2e.yml (hunk 2 above)
Comment (Madhur97) [reply .github/workflows/e2e.yml]: Okay. Minor version upgrade worked in e2e test workflow so this test has been from this workflow. This test can be run with other PRs after #584 is completed.

File: test/e2e/cluster_upgrade_test.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	})
+
+	It("Should successfully upgrade cluster", func() {
+		clusterName := fmt.Sprintf("%s-%s", specName, util.RandomString(6))
+
+		dClient, err := client.NewClientWithOpts(client.FromEnv)
+		dockerClient = dClient
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Creating byohost capacity pool containing 4 hosts")
```
Comment (dharmjit) [inline test/e2e/cluster_upgrade_test.go:64]:
```suggestion
		By("Creating byohost capacity pool containing 4 docker hosts")
```

**Review by dharmjit — APPROVED** [review]

**Review by anusha94 — APPROVED** [review]
lgtm

**Pull request merged**